barretenberg/cpp/pil/vm2/poseidon2_hash.pil

```diff
@@ -4,86 +4,91 @@ include "poseidon2_perm.pil";
-    end * (1 - end) = 0;
-    // The final result of the output of the hash should match the output from the last permutation (b_0)
-    end * (output - b_0) = 0;
-    pol commit num_perm_rounds_rem_inv;
-    // end == 1 when the num_perm_rounds_rem == 0
-    sel * (num_perm_rounds_rem * (end * (1 - num_perm_rounds_rem_inv) + num_perm_rounds_rem_inv) - 1 + end) = 0;
-
-    // We perform the "squeeze" / perm operation until end
-    pol commit execute_perm;
-    execute_perm * (1 - execute_perm) = 0;
-    // The squeeze and end selector must be mutually exclusive
-    sel * (1 - end - execute_perm) = 0;
-    // Need an additional helper that holds the inverse of the num_perm_rounds_rem;
+    // Check that the PADDED_LEN = num_rounds * 3 at the start of the computation
+    sel * start * (num_perm_rounds_rem * 3 - PADDED_LEN) = 0;
     // If we still have rounds to perform, the num_perm_rounds_rem is decremented
-    execute_perm * (num_perm_rounds_rem' - num_perm_rounds_rem + 1) = 0;
-
+    sel * (1 - LATCH_CONDITION) * (num_perm_rounds_rem' - num_perm_rounds_rem + 1) = 0;
+    // Need an additional helper that holds the inverse of the num_perm_rounds_rem;
+    pol commit num_perm_rounds_rem_inv;
+    pol NEXT_ROUND_COUNT = num_perm_rounds_rem - 1;
+    // end == 1 when the (num_perm_rounds_rem - 1) == 0
+    sel * (NEXT_ROUND_COUNT * (end * (1 - num_perm_rounds_rem_inv) + num_perm_rounds_rem_inv) - 1 + end) = 0;
 
-    // The input values are represented by a_0, a_1, a_2, a_3
+    //=== Output Chaining ===
+    // The permutation input values are represented by a_0, a_1, a_2, a_3
+    // The permutation output values are represented by b_0, b_1, b_2, b_3
     // This most definitely could be simplified to a lower degree check
-    // the next perm input is constrained to be the previous perm output + the new values to be hashed.
-    // This occurs when we execute_perm = 1 and we are not the start or the end of the poseidon perm
-    pol NEXT_INPUT_IS_PREV_OUTPUT_SEL = execute_perm' * (1 - start) * execute_perm;
+    // The next perm input is constrained to be the previous perm output + the new values to be hashed if we
+    // are not at the start or the end of hashing
+    pol NEXT_INPUT_IS_PREV_OUTPUT_SEL = (1 - LATCH_CONDITION) * (1 - start);
     pol commit a_0;
-    start * (a_0 - input_0) = 0;
-    sel * NEXT_INPUT_IS_PREV_OUTPUT_SEL * (a_0' - b_0 - input_0') = 0;
     pol commit a_1;
-    start * (a_1 - input_1) = 0;
-    sel * NEXT_INPUT_IS_PREV_OUTPUT_SEL * (a_1' - b_1 - input_1') = 0;
     pol commit a_2;
-    start * (a_2 - input_2) = 0;
-    sel * NEXT_INPUT_IS_PREV_OUTPUT_SEL * (a_2' - b_2 - input_2') = 0;
     pol commit a_3;
-    start * (a_3 - IV) = 0; // IV is placed in the last slot if this is the start
+
+    sel * start * (a_0 - input_0) = 0;
+    sel * NEXT_INPUT_IS_PREV_OUTPUT_SEL * (a_0' - b_0 - input_0') = 0;
+    sel * start * (a_1 - input_1) = 0;
+    sel * NEXT_INPUT_IS_PREV_OUTPUT_SEL * (a_1' - b_1 - input_1') = 0;
```

Comment:
`sel` is added here as a precaution for the `skippable`. It doesnt increase the degree past max so is fine for now imo